asserters.go

```diff
@@ -74,3 +55,62 @@ func checkLowerBoundary[T Type, T2 Type](value T, boundary T2) error {
 
 	return nil
 }
+
+func isFloatOverflow[T Type, T2 Type](value T, boundary T2) bool {
+	// boundary is positive when checking for an overflow
+
+	// everything fits in float64 without overflow.
+	v := float64(value)
+	b := float64(boundary)
+
+	if v > b*1.01 {
+		// way greater than the maximum value
+		return true
+	}
+
+	if v < b*0.99 {
+		// we are way below the maximum value
+		return false
+	}
+	// we are close to the maximum value
+
+	// let's try to create the overflow
+	// by converting back and forth with type juggling
+	conv := float64(T(T2(v)))
+
+	// the number was between 0.99 and 1.01 of the maximum value
+	// once converted back and forth, we need to check if the value is in the same range
+	// if not, so it's an overflow
+	return conv <= b*0.99
+}
+
+func isFloatUnderOverflow[T Type, T2 Type](value T, boundary T2) bool {
+	// everything fits in float64 without overflow.
+	v := float64(value)
+	b := float64(boundary)
+
+	if b == 0 {
+		// boundary is 0
+		// we can check easily
+		return value < 0
+	}
+
+	if v < b*1.01 { // please note value and boundary are negative here
+		// way below than the minimum value, it would underflow
+		return true
+	}
+
+	if v > b*0.99 { // please note value and boundary are negative here
+		// way greater than the minimum value
+		return false
+	}
+
+	// we are just above to the minimum value
+	// let's try to create the underflow
+	conv := float64(T(T2(v)))
+
+	// the number was between 0.99 and 1.01 of the minimum value
+	// once converted back and forth, we need to check if the value is in the same range
+	// if not, so it's an underflow
+	return conv >= b*0.99
```

Comment:
that can't be correct even with a lot more nines?

I suggest you write a function to find how many mantisse  bits there are in float64 (the answer is 53 but it's better to "prove" it) and float32 and use that as boundary tests (tests, not for runtime, imo, the runtime is covered by the roundtrip conversion, afaik)

Reply:
I wrote such function in https://github.com/fortio/safecast/pull/8

or you can hardcode 53 and 24 bits as defined by the IEEE standard.